tests/test_kohya_loras_scaffold.py

```diff
@@ -0,0 +1,347 @@
+#
+#
+# TODO: REMOVE THIS FILE
+# This file is intended to be used for initial development of new features.
+#
+#
+
+import math
+
+import safetensors
+import torch
+from PIL import Image
+
+from diffusers import DiffusionPipeline, DPMSolverMultistepScheduler, StableDiffusionPipeline
+
+
+# modified from https://github.com/kohya-ss/sd-scripts/blob/ad5f318d066c52e5b27306b399bc87e41f2eef2b/networks/lora.py#L17
+class LoRAModule(torch.nn.Module):
+    def __init__(self, org_module: torch.nn.Module, lora_dim=4, alpha=1.0, multiplier=1.0):
+        """if alpha == 0 or None, alpha is rank (no scaling)."""
+        super().__init__()
+
+        if org_module.__class__.__name__ == "Conv2d":
+            in_dim = org_module.in_channels
+            out_dim = org_module.out_channels
+        else:
+            in_dim = org_module.in_features
+            out_dim = org_module.out_features
+
+        self.lora_dim = lora_dim
+
+        if org_module.__class__.__name__ == "Conv2d":
+            kernel_size = org_module.kernel_size
+            stride = org_module.stride
+            padding = org_module.padding
+            self.lora_down = torch.nn.Conv2d(in_dim, self.lora_dim, kernel_size, stride, padding, bias=False)
+            self.lora_up = torch.nn.Conv2d(self.lora_dim, out_dim, (1, 1), (1, 1), bias=False)
+        else:
+            self.lora_down = torch.nn.Linear(in_dim, self.lora_dim, bias=False)
+            self.lora_up = torch.nn.Linear(self.lora_dim, out_dim, bias=False)
+
+        if alpha is None or alpha == 0:
+            self.alpha = self.lora_dim
+        else:
+            if type(alpha) == torch.Tensor:
+                alpha = alpha.detach().float().numpy()  # without casting, bf16 causes error
+            self.register_buffer("alpha", torch.tensor(alpha))  # Treatable as a constant.
+
+        # same as microsoft's
+        torch.nn.init.kaiming_uniform_(self.lora_down.weight, a=math.sqrt(5))
+        torch.nn.init.zeros_(self.lora_up.weight)
+
+        self.multiplier = multiplier
+
+    def forward(self, x):
+        scale = self.alpha / self.lora_dim
+        return self.multiplier * scale * self.lora_up(self.lora_down(x))
```

Comment:
Questions:

* If we check here
  
 
 https://github.com/huggingface/diffusers/blob/29b1325a5ae28fa8d7f459b372582287ffc571e5/src/diffusers/models/attention_processor.py#L531

  we can notice that the computation of `scale` is different. It's a runtime argument, rather. So, if I am reading it correctly, to be able to get to an `scale=1` (which is the default in Diffusers LoRA), it's the `self.alpha` and `self.lora_dim` that need to be set accordingly. 

* Why is there an additional `multipler`, though?

Reply:
Hmm cool! Fair enough. Thanks for explaining!